### PR TITLE
UIP-2411: Basic DDC Support

### PR DESCRIPTION
**Pull request by clairesarsam-wf**
# Ultimate problem

Consumers need `react-dart` to be compilable with the ddc.

# Solution

* Address ddc runtime errors that were not resolved as part of the strong mode changes.
* Add a new test page that renders both dom and custom components with no children, a list of children, and variadic children (currently commented out due to https://github.com/dart-lang/sdk/issues/29904).

# How to +10/QA
* Verify tests pass locally in Chrome (dart2js) and content_shell
    * `pub run test -p chrome`
    * `pub run test -p content-shell`
* Look at each of the test pages. None should have errors other than pre-existing key errors, expected key errors in `call_and_nosuchmethod_test.html` and an input related error in `react_test.html`.

---

please review: @greglittlefield-wf @jacehensley-wf @aaronlademann-wf

### CODE REVIEW

File: example/test/call_and_nosuchmethod_test.dart [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        customComponent({'key': 'singleItemInList2'}, [customComponent({})]), // This should produce a key warning
+        customComponent({'key': 'twoItemsInList2'}, [customComponent({}), customComponent({})]), // This should produce a key warning
+        customComponent({'key': 'oneVariadicChild2'}, customComponent({'key': '1'})),
+
+        // These tests of variadic children won't pass in the ddc until https://github.com/dart-lang/sdk/issues/29904
+        // is resolved.
+        // customComponent({'key': 'twoVariadicChildren2'}, customComponent({}), customComponent({})),
+        // customComponent({'key': 'fiveVariadicChildren2'}, '', customComponent({}), '', customComponent({}), ''),
+      ])
+  , querySelector('#content'));
```
Comment (greglittlefield-wf) [inline example/test/call_and_nosuchmethod_test.dart]: In order to ensure there are no regressions in variadic children in the VM or dart2js, can we perform these validations in actual test files?

Actually, I have some commits where there are tests wired up for passing in children... lemme see if I can get those changes in a good state for this PR.

Comment (clairesarsam-wf) [inline example/test/call_and_nosuchmethod_test.dart]: Factory tests are now up on this PR. Should I remove this example file? It was a bit useful in poking around ddc runtime errors would be trivial to re-create for that.

File: test/factory/common_factory_tests.dart [hunk 2]
```diff
@@ -39,6 +39,11 @@ void commonFactoryTests(Function factory) {
       var instance = factory({}, ['one', 'two',]);
       expect(getJsChildren(instance), equals(['one', 'two']));
     });
+
+    test('an empty List', () {
+      var instance = factory({}, []);
+      expect(getJsChildren(instance), equals([]));
+    });
```
Comment (greglittlefield-wf) [inline test/factory/common_factory_tests.dart]: Could you add one more test that a single iterable child is converted to a list successfully? (A regression test for your fix in 3dce4a543a113cd86d50774b7ddcbef79f79fec8)

File: example/test/call_and_nosuchmethod_test.dart (hunk 1 above)
Comment (greglittlefield-wf) [inline example/test/call_and_nosuchmethod_test.dart]: I'm fine leaving this example in

File: test/factory/common_factory_tests.dart (hunk 2 above)
Comment (clairesarsam-wf) [inline test/factory/common_factory_tests.dart]: Added in 3153b0b

**Comment by greglittlefield-wf** [comment]
+1

**Comment by aaronlademann-wf** [comment]
+10

**Comment by aaronlademann-wf** [comment]
QA +10

- [x] Testing instruction
    - Verified all test pages have no new errors
- [x] Dev +1's
- [x] Unit tests created/updated
- [x] All unit tests pass
    - Tests pass on content-shell and chrome using Dart 1.23.0

**Pull request merged**